### ISSUE TITLE
ShapeOptimizer and TransposeOptimizer produces wrong graphs.

### ISSUE THREAD

**Issue by xadupre** [onnx/tensorflow-onnx#1665]
**Describe the bug**

Attached file: a non-optimized graphs where TransposeOptimizer (_slice_handler) and ReshapeOptimizer.

[debug_noopt.zip](https://github.com/onnx/tensorflow-onnx/files/6969220/debug_noopt.zip)

**TransposeOptimizer**

It fails at that particular node Transpose:

![image](https://user-images.githubusercontent.com/22452781/129049909-1c587c2a-ee1a-4afb-81c7-b048633b0588.png)

onnxruntime can load the non optimized graph (equal to ``numpy.fft.fft2(numpy.random.rand(3, 1, 4).astype(numpy.float32), (1, 4))``. After optimization, it returns the following error:

```
FAIL : Node (Ma_MatMul__358) Op (MatMul) [ShapeInferenceError] Incompatible dimensions for matrix multiplication
```

**ReshapeOptimizer**

It produces the following error stack:

```
  File "tensorflow-onnx\tests\test_backend.py", line 179, in _run_test_case
    return self.run_test_case(func, feed_dict, [], output_names_with_port, **kwargs)
  File "tensorflow-onnx\tests\backend_test_base.py", line 403, in run_test_case
    self.assert_shapes_correct(g, self.config.allow_missing_shapes, not self.config.skip_onnx_checker)
  File "tensorflow-onnx\tests\backend_test_base.py", line 326, in assert_shapes_correct
    self.assertTrue(tf2onnx_shape is not None)
  File "lib\unittest\case.py", line 682, in assertTrue
    raise self.failureException(msg)
```

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Tensorflow Version: 2.5
- Python version: 3.9
- Opset 13

**Comment by TomWildenhain-Microsoft** [comment]
@xadupre did you double check that this error occurs using the `tools/onnx-optimize.py` script?  I tried and could not reproduce the error.  The shape info might differ if you got these originally from a rfft unit test.

**Comment by xadupre** [comment]
I got it from a unit test. I'll try with the optimizer tool.

**Comment by xadupre** [comment]
I fixed the bug in PR #1649. In two cases, the shape of a node was empty and the code did not expect that.
